node/staking.go

```diff
@@ -20,56 +22,38 @@ import (
-	DepositFuncSignature  = "0xd0e30db0"
-	withdrawFuncSignature = "0x2e1a7d4d"
-	funcSingatureBytes    = 4
+	funcSingatureBytes = 4
+	lockPeriodInEpochs = 3 // This should be in sync with contracts/StakeLockContract.sol
 )
 
-// UpdateStakingList updates the stakes of every node.
-// TODO: read directly from smart contract, or at least check the receipt also for incompleted transaction.
-func (node *Node) UpdateStakingList(block *types.Block) error {
-	signerType := types.HomesteadSigner{}
-	txns := block.Transactions()
-	for i := range txns {
-		txn := txns[i]
-		toAddress := txn.To()
-		if toAddress != nil && *toAddress != node.StakingContractAddress { //Not a address aimed at the staking contract.
-			utils.GetLogInstance().Info("Mismatched Staking Contract Address", "expected", node.StakingContractAddress.Hex(), "got", toAddress.Hex())
-			continue
-		}
-		currentSender, _ := types.Sender(signerType, txn)
-		_, isPresent := node.CurrentStakes[currentSender]
-		data := txn.Data()
-		switch funcSignature := decodeFuncSign(data); funcSignature {
-		case DepositFuncSignature: //deposit, currently: 0xd0e30db0
-			amount := txn.Value()
-			value := amount.Int64()
-			if isPresent {
-				//This means the node has increased its stake.
-				node.CurrentStakes[currentSender] += value
-			} else {
-				//This means its a new node that is staking the first time.
-				node.CurrentStakes[currentSender] = value
+// StakeInfo is the struct for the return value of listLockedAddresses func in stake contract.
+type StakeInfo struct {
+	LockedAddresses  []common.Address
+	BlockNums        []*big.Int
+	LockPeriodCounts []*big.Int
+	Amounts          []*big.Int
+}
+
+// UpdateStakingList updates staking information by querying the staking smart contract.
+func (node *Node) UpdateStakingList(stakeInfo *StakeInfo) {
+	node.CurrentStakes = make(map[common.Address]*big.Int)
+	if stakeInfo != nil {
+		for i, addr := range stakeInfo.LockedAddresses {
+			blockNum := stakeInfo.BlockNums[i]
+			lockPeriodCount := stakeInfo.LockPeriodCounts[i]
+			amount := stakeInfo.Amounts[i]
+
+			startEpoch := core.GetEpochFromBlockNumber(blockNum.Uint64())
+			curEpoch := core.GetEpochFromBlockNumber(node.blockchain.CurrentBlock().NumberU64())
+
+			if startEpoch == curEpoch {
+				continue // The token are counted into stakes at the beginning of next epoch.
 			}
-		case withdrawFuncSignature: //withdraw, currently: 0x2e1a7d4d
-			value := decodeStakeCall(data)
-			if isPresent {
-				if node.CurrentStakes[currentSender] > value {
-					node.CurrentStakes[currentSender] -= value
-				} else if node.CurrentStakes[currentSender] == value {
-					delete(node.CurrentStakes, currentSender)
-				} else {
-					continue //Overdraft protection.
-				}
-			} else {
-				continue //no-op: a node that is not staked cannot withdraw stake.
+			if curEpoch-startEpoch <= lockPeriodCount.Uint64()*lockPeriodInEpochs {
+				node.CurrentStakes[addr] = amount
```

Comment:
Is lockPeriodCount a unit of lockPeriod? e.g. If node will have a lockPeriod = "2".  Then it will be locked for 2*3(=lockPeriodInEpochs) = 6 epochs? If so, this calculation makes sense.

Reply:
yes.